### PR TITLE
Conversion from typed to unstructured should set GVK

### PR DESCRIPTION
**Pull request by ash2k**
**What this PR does / why we need it**:
By convention typed objects do not have `TypeMeta` information filled. When the scheme converts a typed object into Unstructured it should set type meta information explicitly otherwise Unstructured object is not recognizable.

**Special notes for your reviewer**:
Bug was not visible to tests because tests were working with a typed object with `TypeMeta` information filled. It was filled by previous tests. I have refactored tests for better isolation. Also fixed some other issues in tests.

Add `?w=1` to the PR's URL to see the diff with whitespace changes ignored.

**Release note**:
```release-note
NONE
```
/kind bug
/sig api-machinery
/cc @smarterclayton @deads2k

### CODE REVIEW

**Comment by ash2k** [comment]
I always wanted to ask - what is the reason for the convention to clear `TypeMeta` in typed objects? That makes `GetObjectKind() schema.ObjectKind` method in `runtime.Object` unusable unless you know you are working with an Unstructured. I.e. writing generic code that works with `runtime.Object` is harder than necessary. Need to use scheme to get the GVK for it. Or alternatively (this is what I do) you can always set GVK (on a copy) when you get an object from informer.

**Comment by deads2k** [comment]
> I always wanted to ask - what is the reason for the convention to clear TypeMeta in typed objects?

As golang objects, the objects have no native typemeta information until they are serialized.  You can see it most clearly in cases where the same golang struct spans multiple groups and versions as the external types are moved.  apps and extensions come to mind.

Any typemeta information would only be partially correct at best.  That's why the scheme method actually considers multiple objectkinds to be valid.

File: staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go [hunk 1]
```diff
@@ -431,6 +431,7 @@ func (s *Scheme) Convert(in, out interface{}, context interface{}) error {
 				return err
 			}
 			unstructuredOut.SetUnstructuredContent(content)
+			unstructuredOut.GetObjectKind().SetGroupVersionKind(gvk)
```
Comment (jennybuckley) [inline staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: Is there a reason why it isn't necessary to also set the GroupVersionKind here? https://github.com/atlassian/kubernetes/blob/cd8fd313bb4323d58e2d0965aa161267e2c2cd46/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L452-L453

Comment (ash2k) [reply staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go]: There is a test that verifies it all works as expected (internal to unstructured) and it does (I just verified). So it is not necessary. The GVK fields that are picked up by the unstructured converter are set here
https://github.com/atlassian/kubernetes/blob/cd8fd313bb4323d58e2d0965aa161267e2c2cd46/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L576

However I agree that it probably should not do that - if my reading of the code is correct, it sets the GVK for typed objects too, which it probably should not do.
@deads2k maybe you could clarify? Also there is some code above that calls `copyAndSetTargetKind()` - why does it need to set the GVK?

**Comment by ash2k** [comment]
ping @jennybuckley

**Comment by ash2k** [comment]
@kubernetes/sig-api-machinery-pr-reviews can we please merge this while it is not code freeze yet? Seem to be a non-controversial PR.

**Comment by jennybuckley** [comment]
/lgtm
I don't think I can approve though

**Pull request merged**